### ISSUE TITLE
df2genind issues false ploidy warning for substring matched loci names

### ISSUE THREAD

**Issue by emstrick** [thibautjombart/adegenet#218]
Hello,

using the data import function df2genind() the following warning was triggered:

> Warning in df2genind(X = X, pop = as.character(pop), ploidy = 2, ncode = ncode,  :
> The observed allele dosage (0-2006) does not match the defined ploidy (2-2).
> Please check that your input parameters (ncode, sep) are correct.

I checked the code: in function df2genind(), line 181 (relative to function head) the calculation of ploidmat seems to be wrong if locus names occur as sub strings of each other.  The problematic part in the code is:

    ploidmat <- vapply(loc.names, function(i) {
        rowSums(out[, grepl(paste0(i, "\\."), colnames(out)),
            drop = FALSE], na.rm = TRUE)
    }, FUN.VALUE = double(nrow(out)))

For example, we have a locus name
Browse[1]> loc.names[1]
[1] "12_92"

But we have super-string locus names as well 29012_92, 44912_92, 47512_92. This issues false matches:
colnames(out[,which(grepl(paste0(loc.names[1], "\\."),colnames(out)))])
[1] "12_92.04"    "12_92.03"    "29012_92.04" "29012_92.02" "44912_92.03"
[6] "44912_92.01" "47512_92.03" "47512_92.04"

out[,which(grepl(paste0(loc.names[1], "\\."),colnames(out)))]

```
      12_92.04 12_92.03 29012_92.04 29012_92.02 44912_92.03 44912_92.01 47512_92.03 47512_92.04
N1-10        2        0          NA          NA          NA          NA          2           0
N1-11       NA       NA           2           0           2           0         NA          NA
N1-12       NA       NA           2           0           0           2         NA          NA
N1-14       NA       NA           2           0          NA          NA         NA          NA
N1-15        2        0           2           0          NA          NA          2           0
N1-4         2        0           2           0           2           0         NA          NA
...
```

Thus, summation over rows leads to wrong ploidy values  (>2) for individual loci.

Fortunately, as far as I can see, ploidmat is only used for sanity checking against a given ploidy (or am I wrong). Since I don't know if other parts in the code use similar string matching mechanisms, I just wanted to this point out as  potential source of problems.

Fortunately too, a simple work around would be to generate sufficiently unique random alias strings of the same length as loci names...

**Comment by zkamvar** [comment]
Hi @emstrick,

Thank you for pointing this out! I added this very recently and forgot about the fact that partial matches can occur 🤦‍♂️.

This can be fixed by adding `"^"` to the beginning of the `paste0()` command so that the search string becomes "^12_92\\.", forcing an exact match.

**Comment by zkamvar** [comment]
Also, thank you for digging into the code and pinpointing the issue!

**Comment by zkamvar** [comment]
I've made the fix on d730ee4e670bc6f07426633b1cb6891fc2c0c142

You can test to see if this works by installing the "exact-match" branch like so:

```r
devtools::install_github("thibautjombart/adegenet@exact-match")
```

**Comment by emstrick** [comment]
Thanks your fast reply and for the "^" suggestion in the regexp!
